packages/wasm-tests/src/lib.rs

```diff
@@ -0,0 +1,57 @@
+extern crate alloc;
+use fuels_abigen_macro::wasm_abigen;
+
+wasm_abigen!(
+    no_name,
+    r#"[
+        {
+            "type":"contract",
+            "inputs":[
+                {
+                    "name":"SomeEvent",
+                    "type":"struct SomeEvent",
+                    "components": [
+                        {
+                            "name": "id",
+                            "type": "u64"
+                        },
+                        {
+                            "name": "account",
+                            "type": "b256"
+                        }
+                    ]
+                },
+                {
+                    "name":"AnotherEvent",
+                    "type":"struct AnotherEvent",
+                    "components": [
+                        {
+                            "name": "id",
+                            "type": "u64"
+                        },
+                        {
+                            "name": "hash",
+                            "type": "b256"
+                        },
+                        {
+                            "name": "bar",
+                            "type": "bool"
+                        }
+                    ]
+                }
+            ],
+            "name":"takes_struct",
+            "outputs":[]
+        }
+    ]
+    "#
+);
+
+pub fn the_fn() {
+    let a_struct = SomeEvent {
+        id: 20,
+        account: Default::default(),
+    };
+
+    println!("It works! {}", a_struct.id);
+}
```

Comment:
When running `cargo test --target wasm32-unknown-unknown` with the following code:

```suggestion
}


#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test() {
        the_fn()
    }
}

```

I encountered a strange error:
```
/home/voxelot/code/fuel/fuels-rs/target/wasm32-unknown-unknown/debug/deps/wasm_tests-b1e4d09cfabc0966.wasm: 1: Syntax error: word unexpected (expecting ")")
error: test failed, to rerun pass '--lib'

```

Reply:
Given fuels-core still largely relies on std, I'm skeptical about whether any of the `fs` methods will actually be implemented or just panic at runtime. Guess it depends the runtime we use the wasm in.

Reply:
Alternately, would using the typescript sdk abigen w/ assemblyscript be a more dev friendly experience for wasm? @digorithm

Reply:
That first error was because cargo test was trying to run WASM as if it were a native executable... so not gonna work. I put in a WASM test harness to run that test.
I think for mac, the only thing needed is a `brew install llvm`, the one shipped by default does not have a WASM target support.

For loading the json abi file, that only happens in the proc macro. Proc macros are actually compiled prior to the crate that's being built, and only needs to be built for the host architecture, not to WASM. The thing that is important here is that the token stream that is output by the proc macro needs to be supported by WASM. All I need are rust type definitions and a way to deserialize a binary blob into those types, and that's the only thing that the ABI gen should output for WASM.

Reply:
Ah that makes sense, thanks for completing the test harness example.

As far as my mac issue, I used brew llvm and had the wasm target working, but it got stuck on secp256k:

```
LLVM ERROR: malformed uleb128, extends past end
```

As far as the proc macro, that is reassuring. It's surprising to me that cargo doesn't load proc-macros from build-deps instead of normal deps, considering they're only run at build time. I wonder if that's because types from the macro deps that are used in the token stream need to be transitively imported?

Reply:
I thought the `.cargo/config.toml` would've handled that? I guess not though.  I'll put in a separate command for it.

For mac are you env vars pointed at the right binary? I had to do this:
```export PATH="/opt/homebrew/opt/llvm/bin:$PATH"
export CC=/opt/homebrew/opt/llvm/bin/clang
export AR=/opt/homebrew/opt/llvm/bin/llvm-ar
```

Though that issue looks like it's building secp256k to wasm? I don't think we'll ever have a need for making signatures in wasm, so maybe it should be an optional dep of fuels-core?

Reply:
Ah nice, these exports fixed the mac issue for me!

Having secp256k be optional in fuels-core would make sense since fuels-rs only uses secp256k for signing. If someone was writing fancy indexer logic involving some cryptography, they could source their own no-std libs for that. Looks like we'd just need to disable default features for `fuel-tx` when importing it within `fuels-core`.

Reply:
Hmm yeah maybe file an issue to investigate this later.

Reply:
Okay, filed it here: #371